service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java

```diff
@@ -0,0 +1,98 @@
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.ccd.client.model.CaseDetails;
+import uk.gov.hmcts.reform.fpl.model.CaseData;
+import uk.gov.hmcts.reform.fpl.model.CaseManagementOrder;
+import uk.gov.hmcts.reform.fpl.model.HearingBooking;
+import uk.gov.hmcts.reform.fpl.model.HearingDateDynamicElement;
+import uk.gov.hmcts.reform.fpl.model.common.Element;
+import uk.gov.hmcts.reform.fpl.model.common.dynamic.DynamicList;
+import uk.gov.hmcts.reform.fpl.model.common.dynamic.DynamicListElement;
+
+import java.time.LocalDate;
+import java.time.format.FormatStyle;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+
+import static java.util.stream.Collectors.toList;
+import static org.apache.commons.lang3.ObjectUtils.isEmpty;
+
+@Service
+public class DraftCMOService {
+    private final ObjectMapper mapper;
+    private final DateFormatterService dateFormatterService;
+
+    @Autowired
+    public DraftCMOService(DateFormatterService dateFormatterService, ObjectMapper mapper) {
+        this.mapper = mapper;
+        this.dateFormatterService = dateFormatterService;
+    }
+
+    public DynamicList getHearingDateDynamicList(CaseDetails caseDetails) {
+        Map<String, Object> data = caseDetails.getData();
+        CaseData caseData = mapper.convertValue(data, CaseData.class);
+
+        List<Element<HearingBooking>> hearingDetails = caseData.getHearingDetails();
+
+        DynamicList hearingDatesDynamic = buildDynamicListFromHearingDetails(hearingDetails);
+
+        if (!isEmpty(caseData.getCaseManagementOrder())) {
+            prePopulateHearingDateSelection(hearingDetails,
+                hearingDatesDynamic,
+                caseData.getCaseManagementOrder());
+        }
+
+        return hearingDatesDynamic;
+    }
+
+    private void prePopulateHearingDateSelection(List<Element<HearingBooking>> hearingDetails,
+                                                 DynamicList hearingDatesDynamic,
+                                                 CaseManagementOrder caseManagementOrder) {
+        UUID hearingDateId = isEmpty(caseManagementOrder) ? null : caseManagementOrder.getId();
+        // There was a previous hearing date therefore we need to remap it
+        String date = hearingDetails.stream()
+            .filter(Objects::nonNull)
+            .filter(element -> element.getId().equals(caseManagementOrder.getId()))
+            .findFirst()
+            .map(element -> formatLocalDateToMediumStyle(element.getValue().getDate()))
+            .orElse("");
+
+        DynamicListElement listElement = DynamicListElement.builder()
+            .label(date)
+            .code(hearingDateId)
+            .build();
+
+        hearingDatesDynamic.setValue(listElement);
+        if (!hearingDatesDynamic.getListItems().contains(listElement)) {
+            hearingDatesDynamic.getListItems().add(listElement);
+        }
+    }
```

Comment:
why would !hearingDatesDynamic.getListItems().contains(listElement) ever happen? Why are we adding a listElement to getListItems?

Reply:
this is that part that adds the extra item to the list when something doesn't exist. I still don't understand this functionality

Reply:
removed this code.